core/aws-core/src/main/java/software/amazon/awssdk/awscore/endpointdiscovery/EndpointDiscoveryRequest.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.awscore.endpointdiscovery;
+
+import java.util.Map;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+@SdkInternalApi
+public final class EndpointDiscoveryRequest
+    implements ToCopyableBuilder<EndpointDiscoveryRequest.Builder, EndpointDiscoveryRequest> {
+
+    private final Optional<String> operationName;
+    private final Optional<Map<String, String>> identifiers;
+    private final Optional<String> cacheKey;
+    private final boolean required;
+
+    private EndpointDiscoveryRequest(BuilderImpl builder) {
+        this.operationName = builder.operationName;
+        this.identifiers = builder.identifiers;
+        this.cacheKey = builder.cacheKey;
+        this.required = builder.required;
+    }
+
+    public Optional<String> operationName() {
+        return operationName;
+    }
+
+    public Optional<Map<String, String>> identifiers() {
```

Comment:
Should we copy and make this  map unmodifiable in constructor?

Reply:
We certainly could but not sure if we need to given this is an internal use only class.